records/2022-12-13.upgrade-database.md

```diff
@@ -0,0 +1,104 @@
+
+**Minor version upgrades**
+
+In contrast, minor version upgrades include only changes that are backward-compatible with existing applications. You can initiate a minor version upgrade manually by modifying your DB instance. Or you can enable the Auto minor version upgrade option when creating or modifying a DB instance. Doing so means that your DB instance is automatically upgraded after Amazon RDS tests and approves the new version. If your PostgreSQL DB instance is using read replicas, you must first upgrade all of the read replicas before upgrading the primary instance. If your DB instance is in a Multi-AZ deployment, then the writer and any standby replicas are upgraded simultaneously. Therefore, your DB instance might not be available until the upgrade is complete. For more details, see [Automatic minor version upgrades for PostgreSQL](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/USER_UpgradeDBInstance.PostgreSQL.html#USER_UpgradeDBInstance.PostgreSQL.Minor). For information about manually performing a minor version upgrade, see [Manually upgrading the engine version](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/USER_UpgradeDBInstance.Upgrading.html#USER_UpgradeDBInstance.Upgrading.Manual).
+
+### 11.17
+
+Likely not risky.
+
+### 12
+
+There are [breaking changes](https://www.postgresql.org/docs/12/release-12.html#id-1.11.6.18.4) from 11 to 12. Unknown if any affect us.
+
+### 13.14
+
+There are also [breaking changes](https://www.postgresql.org/docs/13/release-13.html#id-1.11.6.14.4) from 12 to 13. Unknown if any affect us.
+
+## Upgrade approaches
+
+From [Upgrade your Amazon RDS for PostgreSQL or Amazon Aurora PostgreSQL database, Part 1: Comparing upgrade approaches](https://aws.amazon.com/blogs/database/part-1-upgrade-your-amazon-rds-for-postgresql-database-comparing-upgrade-approaches/) there are a few approaches we can use.
+
+### Option A: In-place upgrade
+This is in some ways the easiest, in that we essentially tell AWS to upgrade the database and it does it. The downside is that the database will be unavailable for most of the upgrade time, and hence Notify will be down. Also, if something goes wrong we would have to restore the database from a backup (and hope that works).
+
+### Option B: Snapshot approach
+In this approach we would make a copy of the database and upgrade that copy, then switch the app to use the new database. The benefit is that if something doesn't work we can quickly switch back to the old database. This approach still requires downtime during the process though, since data added to the old database would not be in the new database.
```

Comment:
we can in theory use the redis url approach and not have the celery workers read anything - this way we just switch on the celery workers after the migration and they then write to the new DB.

Reply:
We might be able to leverage RDS Proxy for such approach:

* https://aws.amazon.com/blogs/database/improving-application-availability-with-amazon-rds-proxy/
* https://catalog.us-east-1.prod.workshops.aws/workshops/098605dc-8eee-4e84-85e9-c5c6c9e43de2/en-US/lab8-fault-tolerance/proxy

> Failover occurs when the primary database instance becomes inaccessible and another instance takes over as the new primary. This disrupts client connections. Failovers can be planned, when they are induced by administrative actions such as a rolling upgrade, or unplanned, when they occur due to failures. In both cases, you want to reduce downtime to minimize client disruption.

> Amazon RDS has multiple high availability options, and RDS Proxy provides failover recovery benefits for each option. The Amazon RDS Multi-AZ option is a primary and standby configuration with synchronous replication. Aurora offers two high availability options—up to 15 asynchronous replicas and multi-master. All Aurora replicas are available for read-only access. You can elect and promote any replica as a new primary for the cluster in the event of failover without any data loss. Aurora multi-master offers multiple active writers.

> With all of these options, when a failover occurs, the client has to detect the connection failure, discover a new primary, and reconnect to it as quickly as possible. With RDS Proxy, the applications can avoid the complexity associated with failovers and experience faster recovery (in as few as 3 seconds). With Aurora, where failovers are faster, DNS propagation delay is the largest contributor to your overall failover time. RDS Proxy actively monitors database instances and automatically connects clients to the right target. It also maintains idle client connections through database failover without dropping them. Idle connections in this context are connections that don’t have outstanding requests.

Reply:
maybe... this is different from the redis migration where in redis data comes in and then leaves, and it's easy to copy over anything that seems stuck. Here data just circles around inside the database, essentially - there's no way to "drain" it as did for redis 🤔 .